src/main/java/io/jenkins/pluginhealth/scoring/probes/Probe.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2022 Jenkins Infra
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.jenkins.pluginhealth.scoring.probes;
+
+import io.jenkins.pluginhealth.scoring.model.Plugin;
+import io.jenkins.pluginhealth.scoring.model.ProbeResult;
+
+/**
+ * Represents the analyze which can be performed on a plugin
+ */
+public abstract class Probe {
+
+    /**
+     * Starts the analyze on a plugin.
+     * Should only be called by the {@link ProbeEngine#run()} method.
+     *
+     * @param plugin the plugin on which to perform the analyze
+     * @return the result of the analyze in a {@link ProbeResult}
+     */
+    public final ProbeResult apply(Plugin plugin) {
+        return doApply(plugin);
+    }
+
+    /**
+     * Perform the analyze on a plugin
+     *
+     * @param plugin the plugin on which the analyze is done
+     * @return a ProbeResult representing the result of the analyze
+     */
+    protected abstract ProbeResult doApply(Plugin plugin);
+
+    protected abstract String key();
+
+    protected boolean requiresRelease() {
```

Comment:
What is our aim for having this boolean field?

Is it to give us an option like: For a given plugin,
- if `requiresRelease` returns `true` that would mean that we won't run the probe on it again if the last time we ran the probe was "after" the last release date of this plugin.
- if it returns `false` that would mean that we don't care if the last time we ran the probe was before or after the release date of this plugin, we will run the probes every single time regardless.

Reply:
Here, I added this field when testing the `SCMLinkValidationProbe`. The link can only be updated when a new release of the plugin is done (because only a new release can update the values in the update-center). And, at the same time, the probe was costly (API rate limitation). So, my reasoning was: when can I prevent in a clever way, the probe from being executed?
The response is: no need to execute the probe if the probe was previously executed on the plugin, and there is no new release of the plugin since the last execution.

Reply:
That's nice, makes sense.